### ISSUE TITLE
[RFC]: Windowed Hidden-State Collection for vLLM Rollouts

### ISSUE THREAD

**Issue by vx120** [vllm-project/vllm#56916]
# RFC: Request-Scoped Hidden-State Capture Windows

## Motivation

Some applications need hidden states for only a bounded range of generated tokens. Capturing and exporting hidden states for an entire generation adds memory traffic, device-to-host copies, and serialization work for tokens the application does not need.

vLLM should support request-scoped, position-aware hidden-state capture so a request can collect a selected window, such as response positions `[1000, 2000)`, without exporting hidden states before or after that window.

## Goals

* Capture hidden states for a fixed token-position window on a per-request basis.
* Support batches whose requests are at different points in their windows.
* Avoid hidden-state export and associated copies outside the requested window.
* Preserve existing behavior and the fast path when capture is not requested.
* Support speculative decoding, tensor/pipeline parallelism, and CUDA Graphs where the relevant execution path supports selective capture.

## Non-goals

* Avoiding the model's normal internal activation computation.
* Returning full-sequence, all-layer activations as a substitute for selective capture.
* Defining application-specific storage or training behavior.

## Proposed API

Add an optional per-request capture configuration. The exact public API shape can follow vLLM's request conventions; the following illustrates the required semantics:

```yaml
hidden_state_capture:
  mode: fixed
  coordinate: response
  start: 1000
  end: 2000
  layer_ids: [1, 8, 16]
```

* `start` is inclusive and `end` is exclusive; `0 <= start < end`.
* `response` is the default coordinate system. A response offset `r` maps to the hidden position used to predict response token `r`; for prompt length `prompt_len`, this is absolute position `prompt_len - 1 + r`.
* `absolute` may be provided for callers that already work with absolute sequence positions.
* The requested layer selection should be explicit and validated by vLLM.
* When the option is absent or disabled, request behavior remains unchanged.

### Per-request capture state

The engine should create immutable capture metadata when it accepts a request and propagate it through scheduling to the worker execution path. It must be associated with the request rather than stored in a process-global switch or passed through unsupported fields in public sampling parameters.

Each request follows this state machine:

```text
DISABLED → ARMED → CAPTURING → FINISHED
```

* `DISABLED`: no capture was requested.
* `ARMED`: capture was requested, but the request has not reached the window.
* `CAPTURING`: scheduled token positions overlap the requested window.
* `FINISHED`: all requested positions have been processed; request capture state and buffers are released.

### Selective export

The runtime should derive selected rows from the active batch's request IDs and token positions. Only rows inside the requested window may be gathered, combined across selected layers, copied to CPU, or returned to the caller. Requests in `ARMED` or `FINISHED` state must not retain or export hidden-state tensors.

Capture state must remain request-local when requests share a batch. The implementation should not register and remove PyTorch hooks on every decode step; the per-request selection should govern export in the normal execution path.

After the final requested position has been processed, capture for that request must finish and release its buffers. Later decode steps must not produce hidden states for it, even if generation continues.

### Output and alignment

For each request with captured rows, return the hidden states together with metadata equivalent to:

```text
hidden_positions
hidden_window_start
hidden_window_end
layer_ids
request_id
```

`hidden_positions` is authoritative. The returned positions may be a subset of the requested window because generation can stop early or execution can fragment the capture. If a request ends before entering its window, return no hidden-state payload for it.

### Execution compatibility

Selective capture must not enable global `output_hidden_states=True` or silently force eager execution. vLLM should validate whether the active model runner and execution mode can provide the required mapping from scheduled rows to request IDs and sequence positions. If a requested combination is unsupported, return a clear capability error rather than silently exporting full-sequence states or ignoring the request.

## Observability

Expose metrics sufficient to measure capture overhead and effectiveness:

```text
vllm/hidden_capture_requests
vllm/hidden_capture_rows_planned
vllm/hidden_capture_rows_captured
vllm/hidden_capture_bytes_copied
vllm/hidden_capture_copy_ms
```

Benchmark decode throughput before, during, and after the capture window. With the same model, batch size, sequence lengths, and speculative-decoding settings, median throughput outside the window should be no more than 3% below the baseline with hidden-state capture disabled.

## Test plan

* Unit tests for response-relative and absolute position conversion, inclusive and exclusive window boundaries, and requests with different prompt lengths.
* Tests for requests ending before, inside, and after the window, and for mixed batches containing `ARMED`, `CAPTURING`, and `FINISHED` requests.
* Tests that returned hidden rows match `hidden_positions` and that no rows are copied or returned after capture finishes.
* Integration tests for request metadata propagation through the serving, scheduler, and worker paths.
* Compatibility tests for eager and CUDA Graph execution, speculative decoding, tensor/pipeline parallelism, prefix caching, and supported hardware backends.

**Comment by HEMU-VERMA** [comment]
Hi @vx120 — this looks like a great fit for _get_prompt_logprobs_dict's existing machinery in gpu_model_runner.py, which already does almost exactly this (per-request num_computed_tokens-relative slicing of hidden_states, with an in-progress buffer on the request object for state spanning multiple steps). I'd like to pick this up.

Plan:
(1) SamplingParams.hidden_state_window: tuple[int,int]|None, response-relative, converted to an absolute window once prompt_token_ids length is known.
(2) A small dependency-free module for the step/window interval math — written and unit tested already (18 tests, incl. a simulated run with variable-size spec-decode steps).
(3) _get_windowed_hidden_states_dict, mirroring _get_prompt_logprobs_dict's shape.
(4) Hook into the three target_hidden_states branches in _execute_model for spec decode — I'd like a sanity check on this part before I build it, since I can't validate it against a real spec-decode run in my current setup.

Open questions: should the window be response-relative (my assumption, per the RFC text) or absolute-from-prompt-start? Should PoolingParams get an equivalent for single-pass embedding use cases?

Happy to open a draft PR with what's built so far and iterate from there, or wait for a design nod first — whichever's preferred.

**Comment by vx120** [comment]
> Hi [@vx120](https://github.com/vx120) — this looks like a great fit for _get_prompt_logprobs_dict's existing machinery in gpu_model_runner.py, which already does almost exactly this (per-request num_computed_tokens-relative slicing of hidden_states, with an in-progress buffer on the request object for state spanning multiple steps). I'd like to pick this up.
> 
> Plan: (1) SamplingParams.hidden_state_window: tuple[int,int]|None, response-relative, converted to an absolute window once prompt_token_ids length is known. (2) A small dependency-free module for the step/window interval math — written and unit tested already (18 tests, incl. a simulated run with variable-size spec-decode steps). (3) _get_windowed_hidden_states_dict, mirroring _get_prompt_logprobs_dict's shape. (4) Hook into the three target_hidden_states branches in _execute_model for spec decode — I'd like a sanity check on this part before I build it, since I can't validate it against a real spec-decode run in my current setup.
> 
> Open questions: should the window be response-relative (my assumption, per the RFC text) or absolute-from-prompt-start? Should PoolingParams get an equivalent for single-pass embedding use cases?
> 
> Happy to open a draft PR with what's built so far and iterate from there, or wait for a design nod first — whichever's preferred.

Thanks for taking such a close look at this and for sharing the implementation plan — the connection to `_get_prompt_logprobs_dict` is particularly helpful.

I’ve already been working on the implementation of this RFC locally, including the request-scoped capture lifecycle and the speculative-decoding integration, and I’d like to continue driving the initial implementation so that it stays aligned with the RFC design.

I’ve now opened the main implementation PR here: https://github.com/vllm-project/vllm/pull/57736

Since #57663 is already open, I think the best way forward is to coordinate rather than develop two parallel implementations. The interval/window math and tests you’ve added look useful, especially for handling variable-size speculative-decoding steps, and I’d be happy to reuse or incorporate those pieces once we align on the final API and runner-side design.

For the API semantics, I’d keep the window response-relative, consistent with the RFC. For the initial scope, I’d also leave `PoolingParams` out and keep the implementation focused on the speculative-decoding rollout path.

For the initial implementation, I’d like to continue owning and driving the end-to-end path through #57736. I expect to complete the main PR within about a week. There should still be plenty of room to collaborate through review, additional testing, and follow-up pieces once the main structure is in place. In particular, your help reviewing the request/window bookkeeping and speculative-decoding paths would be very valuable, and the interval-math tests may fit naturally into that work as well.

Thanks again for jumping in and offering to help.
